Derive StorageAccess for struct

maciejka opened the pull request:
This is an incomplete draft of storage_access derive for structs.

I am stuck at how to compose `StorageAccess` operations. For example for a struct:
```
struct AB {
  a: u256,
  b: u32
}
```
I would like to write it in a series of fields StorageAccess writes:
```
    fn write(address_domain: u32, base: starknet::StorageBaseAddress, value: AB) -> starknet::SyscallResult<()> {
        starknet::StorageAccess::<u256>::write(address_domain, base, value.a)?;
        starknet::StorageAccess::<u32>::write(address_domain, next_address, value.b)?;
        starknet::SyscallResult::Ok(())
    }
```
The problem is that information about space used by previous write is not available in the StorageAccess trait and I see no general way to calculate next_address.  

@spapinistarkware any sugestions?

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/starkware-libs/cairo/3062)
<!-- Reviewable:end -->

spapinistarkware reviewed (commented):
One solution is const generics and exprs, which wedont haev yet and is hard.
Another solution is to ahve associated consts on the StorageAccess trait, which we also don't have yet. This might be a viable path though.
Another solution is to have annotations on the struct fields `#[size(4)]` as a temporary solution. But it won't be verified in any way, and can lead to bugs.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NUuaPfw7YIYBPJoruJJ:b-nddzie)* status: 0 of 5 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

maciejka commented:
Is it a const? Size is a function of value I suppose, e.g.: Option type.
Unless this kind of optimization is not important.

pon., 8 maj 2023 o 11:10 Shahar Papini ***@***.***>
napisał(a):

> ***@***.**** commented on this pull request.
>
> One solution is const generics and exprs, which wedont haev yet and is
> hard.
> Another solution is to ahve associated consts on the StorageAccess trait,
> which we also don't have yet. This might be a viable path though.
> Another solution is to have annotations on the struct fields #[size(4)]
> as a temporary solution. But it won't be verified in any way, and can lead
> to bugs.
>
> *Reviewable
> <https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NUuaPfw7YIYBPJoruJJ:b-nddzie>*
> status: 0 of 5 files reviewed, all discussions resolved
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/starkware-libs/cairo/pull/3062#pullrequestreview-1416430399>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABOTB5MIX3Y5AWPB3CPPC3XFC2ANANCNFSM6AAAAAAXZT7TJM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
mk

spapinistarkware reviewed (commented):
There is no "written in stone" answer here. You can decide how the storage layout works. Yuo have all the ifnormation regarding how StorageAccess works.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NUusKiE7CruAn1agIMD:b4aevex)* status: 0 of 5 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

maciejka commented:
I would either turn `base` argument into a reference:
```
trait StorageAccess<T> {
    fn read(address_domain: u32, ref base: StorageBaseAddress) -> SyscallResult<T>;
    fn write(address_domain: u32, ref base: StorageBaseAddress, value: T) -> SyscallResult<()>;
}
```
or, if we want to save on memory operations in case of single storage operations, add another pair of read/write methods for consecutive operations:
```
trait StorageAccess<T> {
    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<T>;
    fn write(address_domain: u32, base: StorageBaseAddress, value: T) -> SyscallResult<()>;
    fn read_con(address_domain: u32, ref base: StorageBaseAddress) -> SyscallResult<T>;
    fn write_con(address_domain: u32, ref base: StorageBaseAddress, value: T) -> SyscallResult<()>;
}
```
WDYT?

spapinistarkware commented:
> I would either turn `base` argument into a reference:
> 
> ```
> trait StorageAccess<T> {
>     fn read(address_domain: u32, ref base: StorageBaseAddress) -> SyscallResult<T>;
>     fn write(address_domain: u32, ref base: StorageBaseAddress, value: T) -> SyscallResult<()>;
> }
> ```
> 
> or, if we want to save on memory operations in case of single storage operations, add another pair of read/write methods for consecutive operations:
> 
> ```
> trait StorageAccess<T> {
>     fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<T>;
>     fn write(address_domain: u32, base: StorageBaseAddress, value: T) -> SyscallResult<()>;
>     fn read_con(address_domain: u32, ref base: StorageBaseAddress) -> SyscallResult<T>;
>     fn write_con(address_domain: u32, ref base: StorageBaseAddress, value: T) -> SyscallResult<()>;
> }
> ```
> 
> WDYT?

I think making it a ref is a good option.
@orizi ?

orizi reviewed (commented):
this sounds like a good idea - but i think we'd maybe want to update this later somehow - 
so if we can have a worse name for `*_con` variants - that would be even better.
maybe even add `read_con_internal` or something of that manner.

> Reviewed 2 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NV96lP-7yFLdgVS8HxG:btshx31)* status: 2 of 5 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

orizi reviewed (commented):
this lacks some of the meaning `read_continuous_internal` sounds very explicit - but it shouldn't be called directly anyway - so a bad name is good IMO.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NV9K4N45lXmV2A4rgY_:b-mpqcde)* status: 2 of 5 files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

maciejka commented:
Agree. I will proceed with implementation then.

> ***@***.**** commented on this pull request.
>
> this lacks some of the meaning read_continuous_internal sounds very
> explicit - but it shouldn't be called directly anyway - so a bad name is
> good IMO.
>
> *Reviewable
> <https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NV9K4N45lXmV2A4rgY_:b-mpqcde>*
> status: 2 of 5 files reviewed, all discussions resolved
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/starkware-libs/cairo/pull/3062#pullrequestreview-1422260859>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABOTBZMDLSW33W6ZPHQAV3XFS4XTANCNFSM6AAAAAAXZT7TJM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

maciejka commented:
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)
> 
> _[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVVCIy_DrYgjSJ74ShU:-NVVCIy_DrYgjSJ74ShV:bcad4fs) (no related file):_ you should just increase the u8 offset - not the base. the idea of `storage_address_from_base_and_offset(base, offset)` is exactly the fact that we know this never fails, and base is limited to 2^251 - 256.

Fair enough. In this case I would go with straightforward `offset` argument on consecuitive methods and additional  :
```
trait StorageAccess<T> {
    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<T>;
    fn write(address_domain: u32, base: StorageBaseAddress, value: T) -> SyscallResult<()>;
    fn read_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8
    ) -> SyscallResult<T>;
    fn write_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8, value: T
    ) -> SyscallResult<()>;
    fn size_internal(value: T) -> u8;
}
```
with intention to use it like this:
```
impl StorageAccessU256 of StorageAccess<u256> {
    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<u256> {
        StorageAccess::<u256>::read_consecutive_internal(address_domain, base, 0_u8)
    }
    fn write(address_domain: u32, base: StorageBaseAddress, value: u256) -> SyscallResult<()> {
        StorageAccess::<u256>::write_consecutive_internal(address_domain, base, 0_u8, value)
    }
    fn read_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8
    ) -> SyscallResult<u256> {
        let low = StorageAccess::<u128>::read(address_domain, base)?;
        let high = StorageAccess::<u128>::read_consecutive_internal(
            address_domain, base, offset + StorageAccess::<u128>::size_internal(low)
        )?;
        Result::Ok(
            u256 {
                low,
                high
            }
        )
    }
    fn write_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8, value: u256
    ) -> SyscallResult<()> {
        StorageAccess::<u128>::write(address_domain, base, offset, value.low)?;
        StorageAccess::<u128>::write(
            address_domain, base, offset + StorageAccess::<u128>::size_internal(value.low), value.high
        )?;
    }
}
```
WDYT?

orizi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVVCIy_DrYgjSJ74ShU:-NVcBBUs3GfhxApkCHLt:bmciw1b) (no related file):*
<details><summary><i>Previously, maciejka (Maciej Kamiński) wrote…</i></summary><blockquote>

Fair enough. In this case I would go with straightforward `offset` argument on consecuitive methods and additional  :
```
trait StorageAccess<T> {
    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<T>;
    fn write(address_domain: u32, base: StorageBaseAddress, value: T) -> SyscallResult<()>;
    fn read_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8
    ) -> SyscallResult<T>;
    fn write_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8, value: T
    ) -> SyscallResult<()>;
    fn size_internal(value: T) -> u8;
}
```
with intention to use it like this:
```
impl StorageAccessU256 of StorageAccess<u256> {
    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<u256> {
        StorageAccess::<u256>::read_consecutive_internal(address_domain, base, 0_u8)
    }
    fn write(address_domain: u32, base: StorageBaseAddress, value: u256) -> SyscallResult<()> {
        StorageAccess::<u256>::write_consecutive_internal(address_domain, base, 0_u8, value)
    }
    fn read_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8
    ) -> SyscallResult<u256> {
        let low = StorageAccess::<u128>::read(address_domain, base)?;
        let high = StorageAccess::<u128>::read_consecutive_internal(
            address_domain, base, offset + StorageAccess::<u128>::size_internal(low)
        )?;
        Result::Ok(
            u256 {
                low,
                high
            }
        )
    }
    fn write_consecutive_internal(
        address_domain: u32, base: StorageBaseAddress, offset: u8, value: u256
    ) -> SyscallResult<()> {
        StorageAccess::<u128>::write(address_domain, base, offset, value.low)?;
        StorageAccess::<u128>::write(
            address_domain, base, offset + StorageAccess::<u128>::size_internal(value.low), value.high
        )?;
    }
}
```
WDYT?

</blockquote></details>

generally sound good.
`write_consecutive_internal` should probably now be `write_at_offset_internal` - but i'm still now sure if this would be preferable to just adding `ref` the offset value.

___


<!-- Sent from Reviewable.io -->

maciejka commented:
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)
> 
> _[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVVCIy_DrYgjSJ74ShU:-NVcBBUs3GfhxApkCHLt:bmciw1b) (no related file):_
> 
> Previously, maciejka (Maciej Kamiński) wrote…
> generally sound good. `write_consecutive_internal` should probably now be `write_at_offset_internal` - but i'm still now sure if this would be preferable to just adding `ref` the offset value.

Agree on `write_at_offset_internal`. 

Adding `ref` to `offset` is problematic since `offset` has to be increased before it is clear that there will be next operation. Which might result in overflow in case of writing data of lenght == 256. I mean, if we have:
```
fn write_at_offset_internal(
    address_domain: u32, base: StorageBaseAddress, ref offset: u8, value: u256
) -> SyscallResult<()>
```
then what should it do when offset is `255`?

orizi reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 0 of 10 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)

*[a discussion](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVVCIy_DrYgjSJ74ShU:-NVcapNgC_RvXiY_t2ZP:b-w0yspi) (no related file):*
<details><summary><i>Previously, maciejka (Maciej Kamiński) wrote…</i></summary><blockquote>

Agree on `write_at_offset_internal`. 

Adding `ref` to `offset` is problematic since `offset` has to be increased before it is clear that there will be next operation. Which might result in overflow in case of writing data of lenght == 256. I mean, if we have:
```
fn write_at_offset_internal(
    address_domain: u32, base: StorageBaseAddress, ref offset: u8, value: u256
) -> SyscallResult<()>
```
then what should it do when offset is `255`?

</blockquote></details>

got it.

___


<!-- Sent from Reviewable.io -->

orizi reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @maciejka and @spapinistarkware)

___
*[`corelib/src/starknet/storage_access.cairo` line 365 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVhn4K17hszOG4HfEVG:-NVhn4K17hszOG4HfEVH:bmv91fp) ([raw file](https://github.com/starkware-libs/cairo/blob/ee7969d0161e785c6343125563f85b8fa25ad599/corelib/src/starknet/storage_access.cairo#L365)):*
> ```Cairo
> impl StorageAccessClassHash of StorageAccess<ClassHash> {
>     fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<ClassHash> {
>         StorageAccess::<ClassHash>::read_at_offset_internal(address_domain, base, 0_u8)
> ```

don't make existing implementations less efficient just for creating less code.



_Suggestion:_
```Cairo
        Result::Ok(
            StorageAccess::<felt252>::read(address_domain, base)?.try_into().expect('Non ClassHash')
        )
```


<!-- Sent from Reviewable.io -->

maciejka commented:
> > Reviewed all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 0 of 10 files reviewed, 2 unresolved discussions (waiting on @maciejka and @spapinistarkware)
> 
> _[`corelib/src/starknet/storage_access.cairo` line 365 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVhn4K17hszOG4HfEVG:-NVhn4K17hszOG4HfEVH:bmv91fp) ([raw file](https://github.com/starkware-libs/cairo/blob/ee7969d0161e785c6343125563f85b8fa25ad599/corelib/src/starknet/storage_access.cairo#L365)):_
> 
> > ```cairo
> > impl StorageAccessClassHash of StorageAccess<ClassHash> {
> >     fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<ClassHash> {
> >         StorageAccess::<ClassHash>::read_at_offset_internal(address_domain, base, 0_u8)
> > ```
> 
> don't make existing implementations less efficient just for creating less code.
> 
> _Suggestion:_
> 
> ```cairo
>         Result::Ok(
>             StorageAccess::<felt252>::read(address_domain, base)?.try_into().expect('Non ClassHash')
>         )
> ```

I've been thinking about exactly this. I just wanted to make it work first before optimising. 

Another question: What should be the critiria for adding `#[inline(always)]` to StorageAccess `read`/`write` methods. It is not clear to mewhy the `write` methods are usually inlined when `read` are not?

orizi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 0 of 25 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)

___
*[`corelib/src/starknet/storage_access.cairo` line 365 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVhn4K17hszOG4HfEVG:-NVinE75-hLtd7pRwvzH:b-a9rl1w) ([raw file](https://github.com/starkware-libs/cairo/blob/ee7969d0161e785c6343125563f85b8fa25ad599/corelib/src/starknet/storage_access.cairo#L365)):*
<details><summary><i>Previously, maciejka (Maciej Kamiński) wrote…</i></summary><blockquote>

I've been thinking about exactly this. I just wanted to make it work first before optimising. 

Another question: What should be the critiria for adding `#[inline(always)]` to StorageAccess `read`/`write` methods. It is not clear to mewhy the `write` methods are usually inlined when `read` are not? 

</blockquote></details>

i think previously we had an issue with inlining panicable code - but no longer - so both makes sense now.


<!-- Sent from Reviewable.io -->

maciejka commented:
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 0 of 25 files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)
> 
> _[`corelib/src/starknet/storage_access.cairo` line 365 at r3](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NVhn4K17hszOG4HfEVG:-NVinE75-hLtd7pRwvzH:b-a9rl1w) ([raw file](https://github.com/starkware-libs/cairo/blob/ee7969d0161e785c6343125563f85b8fa25ad599/corelib/src/starknet/storage_access.cairo#L365)):_
> 
> Previously, maciejka (Maciej Kamiński) wrote…
> i think previously we had an issue with inlining panicable code - but no longer - so both makes sense now.

Fixed existing implenetations and added `#[inline(always)]` to all read/write functions.

orizi reviewed (changes requested):
> Reviewed 1 of 8 files at r2, 1 of 6 files at r3, 1 of 18 files at r5, 1 of 7 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 4 of 25 files reviewed, 4 unresolved discussions (waiting on @maciejka and @spapinistarkware)

___
*[`crates/cairo-lang-starknet/src/plugin/dispatcher.rs` line 221 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15_4gCQL_6B0-09uX:-NW15_4hAafx2Fmlgr8g:b-l6mwxj) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/dispatcher.rs#L221)):*
> ```XML
>     ));
> 
>     // println!("{}", builder.code);
> ```

remove



_Code quote:_
```XML
    ));

    // println!("{}", builder.code);
```
___
*[`crates/cairo-lang-starknet/src/plugin/events.rs` line 26 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15deZ5aczPqkN_6p9:-NW15deZ5aczPqkN_6pA:b-t5wltg) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L26)):*
> ```XML
>     // if !derive_event_needed(&struct_ast, db) {
>     //     return PluginResult::default();
>     // }
> ```

fix

fix commented code all around.



_Code quote:_
```XML
    // if !derive_event_needed(&struct_ast, db) {
    //     return PluginResult::default();
    // }
```
___
*[`crates/cairo-lang-starknet/src/plugin/events.rs` line 206 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15rZT1gIdMAFiXSGv:-NW15rZT1gIdMAFiXSGw:b9xolpf) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L206)):*
> ```XML
> 
> /// Returns true if the type should be derived as an event.
> pub fn derive_event_needed<T: QueryAttrs>(with_attrs: &T, db: &dyn SyntaxGroup) -> bool {
> ```

fix name to make sure it makes sense to use.



_Code quote:_
```XML
derive_event_needed
```


<!-- Sent from Reviewable.io -->

maciejka commented:
> > Reviewed 1 of 8 files at r2, 1 of 6 files at r3, 1 of 18 files at r5, 1 of 7 files at r6, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 4 of 25 files reviewed, 4 unresolved discussions (waiting on @maciejka and @spapinistarkware)
> 
> _[`crates/cairo-lang-starknet/src/plugin/dispatcher.rs` line 221 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15_4gCQL_6B0-09uX:-NW15_4hAafx2Fmlgr8g:b-l6mwxj) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/dispatcher.rs#L221)):_
> 
> > ```
> >     ));
> > 
> >     // println!("{}", builder.code);
> > ```
> 
> remove
> 
> _Code quote:_
> 
> ```
>     ));
> 
>     // println!("{}", builder.code);
> ```
> 
> _[`crates/cairo-lang-starknet/src/plugin/events.rs` line 26 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15deZ5aczPqkN_6p9:-NW15deZ5aczPqkN_6pA:b-t5wltg) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L26)):_
> 
> > ```
> >     // if !derive_event_needed(&struct_ast, db) {
> >     //     return PluginResult::default();
> >     // }
> > ```
> 
> fix
> 
> fix commented code all around.
> 
> _Code quote:_
> 
> ```
>     // if !derive_event_needed(&struct_ast, db) {
>     //     return PluginResult::default();
>     // }
> ```
> 
> _[`crates/cairo-lang-starknet/src/plugin/events.rs` line 206 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15rZT1gIdMAFiXSGv:-NW15rZT1gIdMAFiXSGw:b9xolpf) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L206)):_
> 
> > ```
> > /// Returns true if the type should be derived as an event.
> > pub fn derive_event_needed<T: QueryAttrs>(with_attrs: &T, db: &dyn SyntaxGroup) -> bool {
> > ```
> 
> fix name to make sure it makes sense to use.
> 
> _Code quote:_
> 
> ```
> derive_event_needed
> ```

Removed unnecesary comments. 
What's wrong with `derive_event_needed`? It returns true if event emission functions should be derived for a type. Name makes sense to me.

orizi reviewed (changes requested):
> Reviewed 1 of 5 files at r1, 1 of 17 files at r4, 1 of 7 files at r6, 4 of 4 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: 10 of 25 files reviewed, 3 unresolved discussions (waiting on @maciejka and @spapinistarkware)

___
*[`corelib/src/starknet/storage_access.cairo` line 97 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eaME8LT_FI_Y05jr:-NW1eaME8LT_FI_Y05js:b-r40msx) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/corelib/src/starknet/storage_access.cairo#L97)):*
> ```Cairo
> 
> impl StorageAccessBool of StorageAccess<bool> {
>     #[inline(always)]
> ```

make this in another pr - to reduce the changes in generation sierra code.



_Code quote:_
```Cairo
#[inline(always)]
```
___
*[`crates/cairo-lang-starknet/src/plugin/events.rs` line 206 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15rZT1gIdMAFiXSGv:-NW1e3Bb1JqB1f44oSNg:b-lvhzc2) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L206)):*
<details><summary><i>Previously, maciejka (Maciej Kamiński) wrote…</i></summary><blockquote>

Removed unnecesary comments. 
What's wrong with `derive_event_needed`? It returns true if event emission functions should be derived for a type. Name makes sense to me.

</blockquote></details>

didn't ee you added a parallel function for the storage access - all good.
___
*[`crates/cairo-lang-starknet/src/plugin/storage_access.rs` line 107 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eJmx769DjNcPDdFq:-NW1eJmx769DjNcPDdFr:b2ni1ar) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/crates/cairo-lang-starknet/src/plugin/storage_access.rs#L107)):*
> ```XML
> }
> 
> pub fn derive_storage_access_needed<T: QueryAttrs>(with_attrs: &T, db: &dyn SyntaxGroup) -> bool {
> ```

doc



_Code quote:_
```XML
derive_storage_access_needed
```


<!-- Sent from Reviewable.io -->

maciejka commented:
> > Reviewed 1 of 5 files at r1, 1 of 17 files at r4, 1 of 7 files at r6, 4 of 4 files at r7, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: 10 of 25 files reviewed, 3 unresolved discussions (waiting on @maciejka and @spapinistarkware)
> 
> _[`corelib/src/starknet/storage_access.cairo` line 97 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eaME8LT_FI_Y05jr:-NW1eaME8LT_FI_Y05js:b-r40msx) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/corelib/src/starknet/storage_access.cairo#L97)):_
> 
> > ```cairo
> > impl StorageAccessBool of StorageAccess<bool> {
> >     #[inline(always)]
> > ```
> 
> make this in another pr - to reduce the changes in generation sierra code.
> 
> _Code quote:_
> 
> ```cairo
> #[inline(always)]
> ```
> 
> _[`crates/cairo-lang-starknet/src/plugin/events.rs` line 206 at r6](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW15rZT1gIdMAFiXSGv:-NW1e3Bb1JqB1f44oSNg:b-lvhzc2) ([raw file](https://github.com/starkware-libs/cairo/blob/bccbd3542aa53fb77ee831de94f952f3fd7855ee/crates/cairo-lang-starknet/src/plugin/events.rs#L206)):_
> 
> Previously, maciejka (Maciej Kamiński) wrote…
> didn't ee you added a parallel function for the storage access - all good.
> 
> _[`crates/cairo-lang-starknet/src/plugin/storage_access.rs` line 107 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eJmx769DjNcPDdFq:-NW1eJmx769DjNcPDdFr:b2ni1ar) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/crates/cairo-lang-starknet/src/plugin/storage_access.rs#L107)):_
> 
> > ```
> > }
> > 
> > pub fn derive_storage_access_needed<T: QueryAttrs>(with_attrs: &T, db: &dyn SyntaxGroup) -> bool {
> > ```
> 
> doc
> 
> _Code quote:_
> 
> ```
> derive_storage_access_needed
> ```

Done. There are still changes in erc20 sierra code. Maybe it is due to this change: https://github.com/starkware-libs/cairo/pull/3062/files#diff-4feb70577299f5836d5feacb9c62326358848bdfe48afcf5de19f51a8663e1f9R313 ? Should I delegate it to a separate pr as well?

maciejka commented:
> > Reviewed 1 of 8 files at r2, 9 of 18 files at r5, 8 of 8 files at r8, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)_ status: all files reviewed, 2 unresolved discussions (waiting on @maciejka and @spapinistarkware)
> 
> _[`crates/cairo-lang-starknet/src/plugin/storage_access.rs` line 107 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eJmx769DjNcPDdFq:-NW1yOSs0rJXZLyO7OKn:b-qorhii) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/crates/cairo-lang-starknet/src/plugin/storage_access.rs#L107)):_
> 
> Previously, maciejka (Maciej Kamiński) wrote…
> you still need the inline of write_at_offset_internal to remove the changes - aside from that it should be fine.

Is it ok now?

orizi reviewed (commented):
> Reviewed 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: all files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)

___
*[`crates/cairo-lang-starknet/src/plugin/storage_access.rs` line 107 at r7](https://reviewable.io/reviews/starkware-libs/cairo/3062#-NW1eJmx769DjNcPDdFq:-NW2CAyeAO9pnU_C3SIc:bbo24j7) ([raw file](https://github.com/starkware-libs/cairo/blob/80786e81f754374babc8696dca40b79d8208222d/crates/cairo-lang-starknet/src/plugin/storage_access.rs#L107)):*
<details><summary><i>Previously, maciejka (Maciej Kamiński) wrote…</i></summary><blockquote>

Is it ok now?
</blockquote></details>

looks good!


<!-- Sent from Reviewable.io -->

orizi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NW2D10k17g57mNs7ial:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @maciejka and @spapinistarkware)



<!-- Sent from Reviewable.io -->

maciejka commented:
Any chance of merging this any time soon? @spapinistarkware

spapinistarkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 5 files at r1, 2 of 8 files at r2, 1 of 17 files at r4, 10 of 18 files at r5, 3 of 4 files at r7, 3 of 8 files at r8, 5 of 5 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NW6pVov2w3kOswvG5oI:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @maciejka)



<!-- Sent from Reviewable.io -->

spapinistarkware reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: all files reviewed, 1 unresolved discussion (waiting on @maciejka)



<!-- Sent from Reviewable.io -->

spapinistarkware reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NW6pwHP5vSvM9YEGaxt:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @maciejka)



<!-- Sent from Reviewable.io -->

orizi reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @maciejka)



<!-- Sent from Reviewable.io -->

milancermak commented:
Why does `size_internal` take a `value` argument? Isn't the type size independent of its value?

orizi reviewed (commented):
since there are options where it would be dependent - for example - if we implement this for `Option<VeryLargeStruct>` - we would probably decide to not write as much in the `None` case.

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/cairo/3062#-:-NW7Umva7p8qP2HPlp1b:b7c5us)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->

milancermak commented:
Not being dependent on the value would also allow for some kind of lazy loading or writing only the dirty values (not sure how that would look like in practice tho).





<!-- Sent from Reviewable.io -->

maciejka commented:
It is not clear to me what is the outcome of recent considerations about `sizeInternal`  being value dependent. Should it be changed somehow? What are the usecases it is not compatible with? It would be better to do it before current version is released...

It seems like there are a few simple tasks left to complete introduction of `#derive(storage_access::StorageAccess)` into the code base:
* use it for starknet builtin types like: `u256`, `EthAddress`
* use it for `Dispatcher`'s generated for `#[abi]`
* add `#[inline]` to `StorageAccess` impls for builtin types
Anything else?
WDYT?

maciejka commented:
> It seems like there are a few simple tasks left to complete introduction of `#derive(storage_access::StorageAccess)` into the code base:
> 
> * use it for starknet builtin types like: `u256`, `EthAddress`
> * use it for `Dispatcher`'s generated for `#[abi]`

Here is a pr deriving StorageAccess for `u256`, `EthAddress`, `Dispatcher`: https://github.com/starkware-libs/cairo/pull/3247

moodysalem commented:
This is a breaking change so I would like it better if it just removed the old methods. If you're optimizing for gas I imagine you're going to write custom implementations, since tx cost should be dominated by posting storage diffs to L1 so you'll want to do maximum packing

Also if the size is constant it should maybe be a generic const parameter. Not sure if this is possible but I would prefer constant size where this is used
